megamek/src/megamek/client/bot/princess/ArtilleryTargetingControl.java

```diff
@@ -0,0 +1,433 @@
+                    
+                    // factor the chance to hit when picking a target - if we've got a spotted hex or an auto-hit hex
+                    // we should prefer to hit that over something that may scatter to who knows where
+                    if(wfi.getProbabilityToHit() > 0) {
+                        damageValue *= wfi.getProbabilityToHit();
+                        
+                        if(damageValue > maxDamage) {
+                            topValuedFireInfos.clear();
+                            maxDamage = damageValue;
+                            topValuedFireInfos.add(wfi);
+                        } else if((damageValue == maxDamage) && (damageValue > 0)) {
+                            topValuedFireInfos.add(wfi);
+                        }
+                    }
+                }
+                
+                // this section is long and obnoxious:
+                // Pick a random fire info out of the ones with the top damage level
+                // Use that to create an artillery attack action, set the action's ammo
+                // then set the fire info's attack action to the created attack action
+                // add the fire info to the firing plan
+                if(!topValuedFireInfos.isEmpty()) {
+                    WeaponFireInfo actualFireInfo = topValuedFireInfos.get(Compute.randomInt(topValuedFireInfos.size()));
+                    ArtilleryAttackAction aaa = (ArtilleryAttackAction) actualFireInfo.buildWeaponAttackAction();
+                    int ammoID = findAmmo(shooter, currentWeapon, game);
+                    aaa.setAmmoId(ammoID);
+                    actualFireInfo.setAction(aaa);
+                    retval.add(actualFireInfo);
+                    retval.setUtility(retval.getUtility() + maxDamage);
+                    owner.sendAmmoChange(shooter.getId(), shooter.getEquipmentNum(currentWeapon), ammoID);
+                }
+            } else if(currentWeapon.getType().hasFlag(WeaponType.F_TAG)) {
+                WeaponFireInfo tagInfo = getTAGInfo(currentWeapon, shooter, game, owner);
+                
+                if(tagInfo != null) {
+                    retval.add(tagInfo);
+                    retval.setUtility(retval.getUtility() + tagInfo.getProbabilityToHit());
+                }
+            }
+        }
+        
+        shooter.setSecondaryFacing(originalFacing);
+        
+        return retval;
+    }
+    
+    /**
+     * Worker function that calculates the shooter's "best" actions that result in a TAG being fired.
+     * @param shooter
+     * @param game
+     * @param owner
+     * @return
+     */
+    private WeaponFireInfo getTAGInfo(Mounted weapon, Entity shooter, IGame game, Princess owner) {
+        WeaponFireInfo retval = null;
+        double hitOdds = 0.0;
+        
+        // pretty simple logic here: take the best shot that you have
+        // if multiple best shots, pick randomly
```

Comment:
Well, it would have picked the last non-zero shot at least. Not the best, but...

Fixed hit odds update. I ditched the random pick idea but forgot to update the comment. The idea was going to be that if a unit had multiple TAG systems, it could fire at multiple targets, but it seems like it's not worth spending the time on it, since I don't see very many units with multiple TAG systems.